subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/AbstractIncompleteSnapshotWithChildrenTest.groovy

```diff
@@ -0,0 +1,417 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot
+
+import org.gradle.internal.file.FileType
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import javax.annotation.Nullable
+import java.util.stream.Collectors
+
+@Unroll
+abstract class AbstractIncompleteSnapshotWithChildrenTest<T extends FileSystemNode> extends Specification {
+
+    abstract T createNodeFromFixture(VirtualFileSystemTestFixture fixture);
+
+    abstract boolean isSameNodeType(FileSystemNode node)
+
+    def "invalidate child with no common pathToParent has no effect (#vfsSpec)"() {
+        def vfsFixture = fixture(vfsSpec)
+        def metadataSnapshot = createNodeFromFixture(vfsFixture)
+        expect:
+        metadataSnapshot.invalidate(vfsFixture.absolutePath, vfsFixture.offset).get() == metadataSnapshot
+
+        where:
+        vfsSpec << NO_COMMON_PREFIX + COMMON_PREFIX
+    }
+
+
+    def "store #fileType child with common prefix adds a new child with the shared prefix of type Directory (#vfsSpec)"() {
+        def vfsFixture = fixture(vfsSpec)
+        def currentNode = createNodeFromFixture(vfsFixture)
+        def newPathToParent = vfsFixture.absolutePath.substring(vfsFixture.offset)
+        def snapshot = mockSnapshot(newPathToParent)
+
+        when:
+        def stored = currentNode.store(vfsFixture.absolutePath, vfsFixture.offset, snapshot)
```

Comment:
I think we need better names here.

- what's "current" about `currentNode`? Maybe it's the "root" node? Or the "parent"? Or maybe even the `grandParent` here?
- again, what is a `vfsFixture`? It seems to embody both the state of the VFS _and_ some information about the test itself. But when I read `vfsFixture.absolutePath` I'm buffled as to what that means.
- `stored` sounds like the name for a `boolean` (it is a past tense verb, not a noun). Maybe `updatedNode`? Or `updatedRoot` (vs. `root` or `rootNode`)?